JoinColony/colonyCDapp
# [Multi-sig] Feat: Reject multisig and cancel motion

## Description

[Related block-ingestor counterpart](https://github.com/JoinColony/block-ingestor/pull/237)

This PR implements the option to reject a multisig motion.

## Testing

You will need 2 browsers, so invite `amy` or whoever you wish to `planex`.
For checking multi sig motion state, use this handy query
```
query MyQuery {
  getColonyAction(
    id: "THE_TX_HASH"
  ) {
    multiSigId
    id
    multiSigData {
      isExecuted
      isSuccess
      isRejected
      nativeMultiSigId
      signatures {
        items {
          multiSigId
          role
          userAddress
          vote
        }
      }
    }
    type
    isMultiSig
  }
}
```

1. Install the multi sig extension and set the global threshold to 2 (easiest to test). (you may need to refresh the app so multi sig pops up as decision method)
2. Assign Amy `Owner` roles with the authority of `Take actions via Multi-Sig`
3. Create a mint tokens action with the decision method set to `Multi-sig`
![image](https://github.com/JoinColony/colonyCDapp/assets/23449297/e98af115-2323-4c8a-8d2a-064999c9f023)
4. Confirm the reject stage of the multisig motion is not visible in the sidebar.

<img width="1065" alt="Screenshot 2024-06-10 at 14 48 03" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/ab55d316-026d-49df-a295-0694e73cc5fb">

4. Copy over the url to the second browser, where you need to be logged in with `amy` (You'll need to have accepted an invite to the colony first).
6. Select `Reject action` from the meatball menu:

<img width="1061" alt="Screenshot 2024-06-10 at 14 47 47" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/aa5113ba-d4fc-48d1-ade3-d16fa3222535">

7. This should vote for reject and open the reject stage of the multisig motion in the sidebar (it will take a few moments for your vote to show - this will be handled more cleanly in the proper UI).

<img width="1062" alt="Screenshot 2024-06-10 at 14 50 12" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/5e0f00eb-0061-4740-8586-55daf56f80c4">

8. Back in the first browser, you should now see the reject stage and Amy's vote. Choose to vote for reject instead of approve.
9. After the threshold for reject has passed, the `Finalize rejection` button should appear below the voting button. You can verify the votes and the state by running the gql query.

<img width="1062" alt="Screenshot 2024-06-10 at 14 52 01" src="https://github.com/JoinColony/colonyCDapp/assets/34915414/f72289ac-2185-4594-bf66-3e06e3a6422c">

10.  Click the `Finalize rejection` button, after some moments, the transaction should be cancelled, the sidebar should change to reflect that the multi sig has been cancelled. You can run the query to make sure :)

![Screenshot 2024-06-07 at 08 35 31](https://github.com/JoinColony/colonyCDapp/assets/34915414/99191a60-e005-4205-ad52-255d14d63fb8)

Further testing:
Create another mint multisig action and try rejecting via the meatball menu as the user who created the action. This should instantly cancel the multisig motion, skipping the entire reject stage (it is also possible to cancel a multisig motion this way even if approve / reject has surpassed the voting threshold but not yet been finalized).

## Diffs

**New stuff** ✨

* Added "reject step" to the multisig widget
* Added new `RejectionChanged` handler to block-ingestor

**Changes** 🏗

* Meatball menu now cancels multisig motion if owner, or sets "reject step" to show if not owner.
* Vote button now accepts a vote type button for approve or reject.
* Refactored some helper functions on block-ingestor to reuse them for both approve and reject

Resolves #2154 and #2158
